[cpp] Null Check Interfaces

Aidan63 opened the pull request:
Adds a null check to the interface object before we start casting, dereferencing, and doing anything which will cause an access violation.

```haxe
interface IFoo {
    function foo():Void;
}

function main() {
    final o : IFoo = null;

    try {
        o.foo();
    catch (exn:Any) {
        trace(exn);
    }

    o.foo();
}
```

With the current behaviour null interface access cannot be caught and you will not get stack traces, adding the null check means it can now be caught and you will get stack traces. The used `NullReference` function is also used by the hxcpp debugger, so null interfaces will no longer crash the debugger.

Aidan63 committed:
Add a IsNull check

Aidan63 committed:
escape quote

Aidan63 committed:
Add new line and remove unneeded dynamic

rainyt commented:
Very nice, i will try it.

rainyt commented:
Is there any way for me to directly download the pre constructed Mac executable file? I encountered difficulties constructing haxe from the code source.

rainyt commented:
OK,  I found it, thank you

rainyt commented:
I cannot directly obtain version 4.3.5, it seems that I can only construct from the source.

Simn commented:
Is there a meaningful test we could add for this?

kLabz commented:
> Is there a meaningful test we could add for this?

Yeah, snippet in PR description segfaults without this fix, and exception can now be caught

Aidan63 committed:
Guard interface null checks behind the same defines as class access

Aidan63 committed:
Add null interface exception test

Aidan63 committed:
move target guard to the right place...

Aidan63 commented:
I've added a test to make sure accessing a null interface does throw.

Also updated the generated code a bit so the null check is guarded behind the `HXCPP_CHECK_POINTER` define. For reference this is the null checking code when accessing class objects.

```C++
      #ifdef HXCPP_CHECK_POINTER
      if (!mPtr) NullReference("Object", true);
      // The handler might have fixed up the null value
      if (!mPtr) NullReference("Object", false);
      #ifdef HXCPP_GC_CHECK_POINTER
         GCCheckPointer(mPtr);
      #endif
      #endif
      return mPtr;
```

I have not done the double NullReference call as that boolean parameter "fixup" isn't used so I'm guessing its a hold over from a old implementation.
